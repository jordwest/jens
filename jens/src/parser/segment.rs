use crate::parser::{get_ident, Rule};
use pest::iterators::Pair;

#[derive(Debug, PartialEq)]
pub enum Segment {
    Content(String),
    Placeholder(String),
}

impl From<Pair<'_, Rule>> for Segment {
    fn from(pair: Pair<'_, Rule>) -> Self {
        match pair.as_rule() {
            Rule::escaped_dollar => Segment::Content("$".into()),
            Rule::not_placeholder => Segment::Content(pair.as_str().into()),
            Rule::placeholder => Segment::Placeholder(get_ident(pair)),
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::parse;

    fn tmpl_line(content: &str) -> String {
        format!("main =\n    {}\n----", content)
    }

    #[test]
    fn placeholder() {
        let templates = parse(&tmpl_line("${x}")).unwrap();
        let segments = &templates[0].lines[0].segments;

        assert_eq!(segments, &[Segment::Placeholder("x".into())]);
    }

    #[test]
    fn raw_content() {
        let templates = parse(&tmpl_line("content")).unwrap();
        let segments = &templates[0].lines[0].segments;

        assert_eq!(segments, &[Segment::Content("content".into())]);
    }

    #[test]
    fn escaped_dollar() {
        let templates = parse(&tmpl_line("\\$")).unwrap();
        let segments = &templates[0].lines[0].segments;

        assert_eq!(segments, &[Segment::Content("$".into())]);
    }

    #[test]
    fn escaped_placeholder() {
        let templates = parse(&tmpl_line("\\${x}")).unwrap();
        let segments = &templates[0].lines[0].segments;

        assert_eq!(
            segments,
            &[Segment::Content("$".into()), Segment::Content("{x}".into())]
        );
    }
}
