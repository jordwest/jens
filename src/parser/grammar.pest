wsp = _{ ( " " | "\t" )* }
significant_whitespace = { ( " " | "\t" )+ }
newline = _{ "\n" }

identifier = { ('A'..'z' | '0'..'9' | "_" )+ }

template_empty_line = { "\n" }
template_content = { (!newline ~ ANY)* }
template_line = { significant_whitespace ~ template_content ~ newline }
template_decl = { identifier ~ wsp ~ "=" ~ wsp }
template_terminator = { "-"+ ~ (newline | EOI) }
template = { template_decl ~ newline ~ (template_line | template_empty_line)* ~ template_terminator }

template_one_liner = { template_decl ~ template_content ~ (newline | EOI) }

file = { SOI ~ (template | newline | template_one_liner)* ~ EOI }

placeholder = { "${" ~ identifier ~ "}" }
escaped_dollar = { "\\$" }
not_placeholder = { (!("${" | "\\$") ~ ANY)+ }
template_phase2 = { SOI ~ (escaped_dollar | not_placeholder | placeholder)* ~ EOI }
