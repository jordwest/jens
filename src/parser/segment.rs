use crate::parser::Rule;
use pest::iterators::Pair;

#[derive(Debug, PartialEq)]
pub(crate) enum Segment {
    Content(String),
    Placeholder(String),
}

impl From<Pair<'_, Rule>> for Segment {
    fn from(pair: Pair<'_, Rule>) -> Self {
        match pair.as_rule() {
            Rule::escaped_dollar => Segment::Content("$".into()),
            Rule::not_placeholder => Segment::Content(pair.as_str().into()),
            Rule::placeholder => Segment::Placeholder(get_ident(pair)),
            _ => unreachable!(),
        }
    }
}

fn get_ident(pair: Pair<'_, Rule>) -> String {
    pair.into_inner().nth(0).unwrap().as_str().into()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::parse;

    fn tmpl_line(content: &str) -> String {
        format!("main =\n    {}\n----", content)
    }

    #[test]
    fn placeholder() {
        let templates = parse(&tmpl_line("${x}")).unwrap();
        let segments = &templates[0].lines[0].content;

        assert_eq!(segments, &[Segment::Placeholder("x".into())]);
    }

    #[test]
    fn content() {
        let templates = parse(&tmpl_line("content")).unwrap();
        let segments = &templates[0].lines[0].content;

        assert_eq!(segments, &[Segment::Content("content".into())]);
    }

    #[test]
    fn escaped_dollar() {
        let templates = parse(&tmpl_line("\\$")).unwrap();
        let segments = &templates[0].lines[0].content;

        assert_eq!(segments, &[Segment::Content("$".into())]);
    }

    #[test]
    fn escaped_placeholder() {
        let templates = parse(&tmpl_line("\\${x}")).unwrap();
        let segments = &templates[0].lines[0].content;

        assert_eq!(
            segments,
            &[Segment::Content("$".into()), Segment::Content("{x}".into())]
        );
    }
}
