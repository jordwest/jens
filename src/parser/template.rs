use crate::parser::{get_ident, parse_phase2, segment::Segment, Rule};
use pest::iterators::Pair;

#[derive(Debug, Default, PartialEq)]
pub(crate) struct Template {
    pub(crate) name: String,
    pub(crate) indent_ignored: usize,
    pub(crate) lines: Vec<TemplateLine>,
}

impl<'a> From<Pair<'a, Rule>> for Template {
    fn from(pair: Pair<'a, Rule>) -> Template {
        let mut template = Template::default();

        for item in pair.into_inner() {
            match item.as_rule() {
                Rule::template_content => template.lines.push(item.as_str().into()),
                Rule::template_decl => template.name = get_ident(item),
                Rule::template_line => template.lines.push(item.into()),
                Rule::template_terminator => {
                    template.indent_ignored = item.as_str().matches("-").count()
                }
                Rule::template_empty_line => template.lines.push(TemplateLine::default()),
                _ => unreachable!(),
            }
        }
        template
    }
}

#[derive(Debug, Default, PartialEq)]
pub(crate) struct TemplateLine {
    pub(crate) indentation: String,
    pub(crate) segments: Vec<Segment>,
}

impl From<&str> for TemplateLine {
    fn from(s: &str) -> TemplateLine {
        TemplateLine {
            indentation: "".into(),
            segments: parse_phase2(s).unwrap(),
        }
    }
}

impl<'a> From<Pair<'a, Rule>> for TemplateLine {
    fn from(pair: Pair<'a, Rule>) -> TemplateLine {
        let mut indentation = String::new();
        let mut segments = vec![];
        for item in pair.into_inner() {
            match item.as_rule() {
                Rule::significant_whitespace => indentation = item.as_str().into(),
                Rule::template_content => segments = parse_phase2(item.as_str()).unwrap(),
                _ => unreachable!(),
            }
        }
        TemplateLine {
            indentation,
            segments,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::{parse, segment::Segment};

    #[test]
    fn empty_template() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n----").unwrap();

        assert_debug_snapshot_matches!("template.empty_template", templates);
    }

    #[test]
    fn space_indented_content() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n    indent4\n     indent5\n----").unwrap();

        assert_debug_snapshot_matches!("template.indented_content", templates);
    }

    #[test]
    fn tab_indented() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n\tindent1\n\t\tindent2\n-").unwrap();

        assert_debug_snapshot_matches!("template.tab_indented", templates);
    }

    #[test]
    fn one_liner() {
        use insta::assert_debug_snapshot_matches;
        let templates =
            parse("main =      this is a one-liner and white space at the beginning is ignored")
                .unwrap();

        assert_debug_snapshot_matches!("template.one_liner", templates);
    }

    #[test]
    fn mixed_indentation() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n    \tindent\n----").unwrap();

        assert_debug_snapshot_matches!("template.mixed_indentation", templates);
    }

    #[test]
    fn ignore_more_indentation() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n  content\n----").unwrap();

        assert_debug_snapshot_matches!("template.ignore_more_indentation", templates);
    }

    #[test]
    fn ignore_inner_template() {
        use insta::assert_debug_snapshot_matches;
        let templates = parse("main =\n    main =\n        x\n    ----\n----").unwrap();

        assert_debug_snapshot_matches!("template.ignore_inner_template", templates);
    }

    #[test]
    fn invalid_start() {
        assert!(parse("main\n  content\n--").is_err())
    }

    #[test]
    fn invalid_end() {
        assert!(parse("main =\n content\n").is_err())
    }
}
