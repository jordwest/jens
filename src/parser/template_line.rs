use crate::parser::{segment::Segment, GrammarParser, Rule};
use pest::{iterators::Pair, Parser};

#[derive(Debug, Default, PartialEq)]
pub(crate) struct TemplateLine {
    pub(crate) indentation: String,
    pub(crate) content: Vec<Segment>,
}

impl<'a> From<Pair<'a, Rule>> for TemplateLine {
    fn from(pair: Pair<'a, Rule>) -> TemplateLine {
        let mut indentation = String::from("");
        let mut content = vec![];
        for item in pair.into_inner() {
            match item.as_rule() {
                Rule::significant_whitespace => indentation = String::from(item.as_str()),
                Rule::template_content => {
                    let result = GrammarParser::parse(Rule::template_phase2, item.as_str())
                        .unwrap()
                        .next()
                        .unwrap();
                    for segment in result.into_inner() {
                        content.push(Segment::from(segment));
                    }
                }
                unknown => panic!("Unexpected rule '{:?}' found", unknown),
            }
        }
        TemplateLine {
            indentation,
            content,
        }
    }
}
