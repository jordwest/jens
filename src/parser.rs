use self::template::Template;
use pest::{self, Parser};

pub(crate) mod template;

mod segment;
mod template_line;

#[cfg(debug_assertions)]
const _GRAMMAR: &'static str = include_str!("grammar.pest"); // relative to this file

#[derive(Parser)]
#[grammar = "grammar.pest"]
struct GrammarParser;

pub(crate) fn parse(content: &str) -> Result<Vec<Template>, pest::Error<Rule>> {
    GrammarParser::parse(Rule::file, content)
        .and_then(|mut pairs| Ok(pairs.next().unwrap()))
        .and_then(|pair| {
            let mut templates: Vec<Template> = vec![];
            for item in pair.into_inner() {
                match item.as_rule() {
                    Rule::template => templates.push(Template::from(item)),
                    unknown => panic!("Unexpected rule '{:?}' found", unknown),
                }
            }
            Ok(templates)
        })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::{segment::Segment, template_line::TemplateLine};

    const TEST_TEMPLATE: &str = "template1 =
    line 1 with ${placeholder} in the middle
----

template2 =
  a line without a placeholder
  but with an \\${escaped} dollar sign
--
";

    #[test]
    fn parses_a_template() {
        assert_eq!(
            parse(TEST_TEMPLATE).unwrap(),
            vec![
                Template {
                    name: "template1".into(),
                    indent_ignored: 4,
                    lines: vec![TemplateLine {
                        indentation: "    ".into(),
                        content: vec![
                            Segment::Content("line 1 with ".into()),
                            Segment::Placeholder("placeholder".into()),
                            Segment::Content(" in the middle".into()),
                        ],
                    },]
                },
                Template {
                    name: "template2".into(),
                    indent_ignored: 2,
                    lines: vec![
                        TemplateLine {
                            indentation: "  ".into(),
                            content: vec![Segment::Content("a line without a placeholder".into())],
                        },
                        TemplateLine {
                            indentation: "  ".into(),
                            content: vec![
                                Segment::Content("but with an ".into()),
                                Segment::Content("$".into()),
                                Segment::Content("{escaped} dollar sign".into()),
                            ]
                        }
                    ]
                }
            ]
        );
    }
}
