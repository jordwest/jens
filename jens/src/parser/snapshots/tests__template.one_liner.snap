---
created: "2019-02-13T08:09:16.624806Z"
creator: insta@0.6.2
source: src/parser/template.rs
expression: templates
---
[
    Template {
        name: "main",
        indent_ignored: 0,
        lines: [
            TemplateLine {
                indentation: "",
                segments: [
                    Content(
                        "this is a one-liner and white space at the beginning is ignored"
                    )
                ]
            }
        ]
    }
]
